python/runtime/db.py

```diff
@@ -25,50 +25,76 @@ def connect_with_data_source(driver_dsn):
+# NOTE(typhoonzero): use a magic number to represent missing values for
+# xgboost, since we can not write a dmatrix file with NaN.
+XGBOOST_NULL_MAGIC = 9999.0
 
 
 def read_feature(raw_val, feature_spec, feature_name):
     # FIXME(typhoonzero): Should use correct dtype here.
     if feature_spec["is_sparse"]:
         if feature_spec["format"] == "kv":
-            items = raw_val.split()
-            items = [item.split(':', 2) for item in items]
-            indices = np.array([int(item[0]) for item in items],
-                               dtype=np.int64)
-            values = np.array([float(item[1]) for item in items],
-                              dtype=np.float32)
+            if raw_val is None:
+                indices = np.array([], dtype=np.int64)
+                values = np.array([], dtype=np.float32)
+            else:
+                items = raw_val.split()
+                items = [item.split(':', 2) for item in items]
+                indices = np.array([int(item[0]) for item in items],
+                                   dtype=np.int64)
+                values = np.array([float(item[1]) for item in items],
+                                  dtype=np.float32)
         else:
-            indices = np.fromstring(raw_val,
-                                    dtype=int,
-                                    sep=feature_spec["delimiter"])
-            indices = indices.reshape(indices.size, 1)
-            values = np.ones([indices.size], dtype=np.int64)
+            if raw_val is None:
+                indices = np.array([], dtype=int)
+                values = np.array([], dtype=np.int64)
+            else:
+                indices = np.fromstring(raw_val,
+                                        dtype=int,
+                                        sep=feature_spec["delimiter"])
+                indices = indices.reshape(indices.size, 1)
+                values = np.ones([indices.size], dtype=np.int64)
 
         dense_shape = np.array(feature_spec["shape"], dtype=np.int64)
         return indices, values, dense_shape
     elif feature_spec["delimiter"] != "":
         # Dense string vector
         if feature_spec["dtype"] == "float32":
-            vec = np.fromstring(raw_val,
-                                dtype=np.float32,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                vec = np.array([XGBOOST_NULL_MAGIC], dtype=np.float32)
+            else:
+                vec = np.fromstring(raw_val,
+                                    dtype=np.float32,
+                                    sep=feature_spec["delimiter"])
         elif feature_spec["dtype"] == "int64":
-            vec = np.fromstring(raw_val,
-                                dtype=np.int64,
-                                sep=feature_spec["delimiter"])
+            if raw_val is None:
+                vec = np.array([int(XGBOOST_NULL_MAGIC)], dtype=np.int64)
```

Comment:
The shape is not matched with the not-none case.
Will there be any offset inconsistency between none/not-none case inside [dump_matrix](https://github.com/sql-machine-learning/sqlflow/blob/c67113ecd3cc3115aac8bf56f0f7439054969de6/python/runtime/xgboost/dataset.py#L149-L155).